Gravatar/MruImageCache.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace Gravatar
+{
+    /// <summary>
+    /// Caches most-recently-used images.
+    /// </summary>
+    /// <remarks>
+    /// Decorates an inner cache, delegating to it as needed.
+    /// <para />
+    /// If an image is available in memory, the inner cache can be bypassed.
+    /// </remarks>
+    public sealed class MruImageCache : IImageCache
+    {
+        private readonly int _cleanAtSize;
+        private readonly int _cleanToSize;
+
+        private readonly ConcurrentDictionary<string, Entry> _entryByFileName = new ConcurrentDictionary<string, Entry>();
```

Comment:
❓ Why not implement this with the following?

```csharp
private readonly Dictionary<string, LinkedListNode<Entry>> _linkedEntryByFileName;
private readonly LinkedList<Entry> _entries;
```

It would simplify size management:

1. When reading an image, move the `LinkedListNode<Entry>` to the front of the list
2. When adding an image, remove the last node if the list has already reached the maximum size

📝 The concurrency benefits of `ConcurrentDictionary` do not outweigh the cache eviction simplicity benefits of the approach I describe here.

Reply:
Added a generic `MruCache<TKey, TValue>` with tests. Using linked list nodes was a nice insight. Thanks again.